### PR TITLE
fix: improve redirect validation

### PR DESCRIPTION
**Pull request by markusahlstrand**
Various fixes and improvements to the url checks

### CODE REVIEW

File: src/types/sql/Connection.ts [hunk 1]
```diff
@@ -1,3 +1,17 @@
...
+  tenantId: z.string(),
+  clientId: z.string(),
+  clientSecret: z.string(),
+  authorizationEndpoint: z.string(),
+  tokenEndpoint: z.string(),
+  createdAt: z.string(),
+  modifiedAt: z.string(),
+});
+
 export interface SqlConnection {
```
Comment (danielgent) [inline src/types/sql/Connection.ts:15]: What about changing this Markus to be 

```
export type SqlConnection = z.infer<typeof SqlConnectionSchema>;
``

File: src/utils/validate-redirect-url.ts [hunk 2]
```diff
@@ -10,18 +15,26 @@ export function validateRedirectUrl(
...
+    // This doesn't work in cloudflare workers for whatever reason
+    // const url = new URL(allowedUrl);
+    const match: RegExpMatchArray | null = allowedUrl.match(urlPattern);
+    if (!match) {
+      throw new Error("Invalid URL");
+    }
+
+    // This replaces * with .* and escapes any other regexes in the string
+    const host = escapeRegExp(match[1]).replace(/\\\*/g, ".*");
+    // This removes any trailing slahes in the path and escapes any other regexes in the string
```
Comment (danielgent) [inline src/utils/validate-redirect-url.ts:28]:
```suggestion
    // This removes any trailing slashes in the path and escapes any other regexes in the string
```

File: test/fixtures/context.ts [hunk 3]
```diff
@@ -33,6 +33,29 @@ interface stateInput {
...
 }
 
+const client: Client = {
+  id: "id",
+  name: "clientName",
+  clientSecret: "clientSecret",
+  tenantId: "tenantId",
+  senderEmail: "senderEmail",
+  senderName: "senderName",
+  allowedCallbackUrls: ["http://localhost:3000", "https://example.com"],
```
Comment (danielgent) [inline test/fixtures/context.ts:43]: You could change these to be `/callback` which would be more realistic *and* we could check that we aren't mixing up any of these as they're all identical

File: test/routes/tsoa/token.spec.ts [hunk 4]
```diff
@@ -26,24 +27,37 @@ describe("token", () => {
...
+    connections: [
+      {
+        name: "google-oauth2",
+        clientId: "googleClientId",
+        clientSecret: "googleClientSecret",
+        authorizationEndpoint: "https://accounts.google.com/o/oauth2/v2/auth",
+        tokenEndpoint: "https://oauth2.googleapis.com/token",
+      },
+    ],
+  };
```
Comment (danielgent) [inline test/routes/tsoa/token.spec.ts:51]: As mentioned, there is definitely an argument for defining these in a common fixtures file and then importing `EXAMPLE_CLIENT_FIXTURE` from there :thinking: 


The disadvantage is one level of indirection for anyone looking at the code

File: test/routes/tsoa/token.spec.ts [hunk 5]
```diff
@@ -172,17 +175,7 @@ describe("token", () => {
...
-            id: "publisherClientId",
-            callbackUrls: ["https://example.com"],
-            scopes: ["profile", "email", "openid"],
-            secrets: [
-              {
-                id: "secretId",
-                hash: "clientSecret",
-              },
-            ],
-          }),
```
Comment (danielgent) [inline test/routes/tsoa/token.spec.ts:185]: Nice so `zod` picked all these up :sunglasses:

File: test/utils/validate-redirect-url.spec.ts [hunk 6]
```diff
@@ -2,64 +2,76 @@ import { validateRedirectUrl } from "../../src/utils/validate-redirect-url";
...
+    const allowedUrls = ["http://example.com"];
+    const redirectUri = "http://example.com/";
+    expect(() => validateRedirectUrl(allowedUrls, redirectUri)).not.toThrow();
+  });
+
+  it("should should ignore query strings", () => {
+    const allowedUrls = ["http://example.com"];
+    const redirectUri = "http://example.com?foo=bar";
+    expect(() => validateRedirectUrl(allowedUrls, redirectUri)).not.toThrow();
+  });
```
Comment (danielgent) [inline test/utils/validate-redirect-url.spec.ts:76]: None of your tests have more than one `allowedUrls` :smile: 

So the code could only work for a lenght of 1, and we'd never know :clown_face:

Comment (danielgent) [reply test/utils/validate-redirect-url.spec.ts:76]: Do we need a few with multiple allowedURLs
* matches first
* matches last
* matches middle
* matches some
* matches none

:shrug:   It's possible to write infinite tests

Comment (danielgent) [reply test/utils/validate-redirect-url.spec.ts:76]: *also* you could use a `describe.each` and feed in a data structure instead. I've got examples of this on some of the repos.  Could be `{input, expectedOutput, description}`

**Review by danielgent — APPROVED** [review]
I definitely have some comments but good to merge :t-rex:

File: src/types/sql/Connection.ts (hunk 1 above)
Comment (markusahlstrand) [reply src/types/sql/Connection.ts:15]: I think TSOA wouldn't like this.. I would love to investigate this further. Another option to zod would be to use the validations we get via TSOA..

**Pull request merged**